server/src/service.rs

```diff
@@ -115,17 +117,41 @@ impl Service {
     #[zbus(out_args("unlocked", "prompt"))]
     pub async fn unlock(
         &mut self,
-        _objects: Vec<OwnedObjectPath>,
-    ) -> Result<(Vec<OwnedObjectPath>, ObjectPath), ServiceError> {
-        todo!()
+        objects: Vec<OwnedObjectPath>,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(Vec<OwnedObjectPath>, OwnedObjectPath), ServiceError> {
+        let unlocked = self.set_locked(false, &objects).await?;
+        if unlocked.is_empty() {
+            let n_prompt = *self.prompt_index.read().await + 1;
+            let prompt = Prompt::new(n_prompt);
+            *self.prompt_index.write().await = n_prompt;
+            let path = prompt.path().clone();
+            object_server.at(path.clone(), prompt).await?;
+
+            return Ok((unlocked, path));
+        }
+
+        Ok((unlocked, OwnedObjectPath::default()))
     }
 
     #[zbus(out_args("locked", "prompt"))]
     pub async fn lock(
         &mut self,
-        _objects: Vec<OwnedObjectPath>,
-    ) -> Result<(Vec<OwnedObjectPath>, ObjectPath), ServiceError> {
-        todo!()
+        objects: Vec<OwnedObjectPath>,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(Vec<OwnedObjectPath>, OwnedObjectPath), ServiceError> {
+        let locked = self.set_locked(true, &objects).await?;
+        if locked.is_empty() {
+            let n_prompt = *self.prompt_index.read().await + 1;
+            let prompt = Prompt::new(n_prompt);
+            *self.prompt_index.write().await = n_prompt;
```

Comment:
If you pass the service manager when creating the prompt, you could deal with incrementing the prompt index there

Reply:
Done, thanks! I think we should also do the same for session :)